File: src/sc_neurocore/accel/julia/neurons/clif.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for clif

module ClifAccel

export step!, simulate, ComplementaryLIFNeuronState, validate

const V_MAX = 1.0e12

mutable struct ComplementaryLIFNeuronState
    v_pos::Float64
    v_neg::Float64
    tau::Float64
    v_threshold::Float64
    dt::Float64
    alpha::Float64
end

function ComplementaryLIFNeuronState()
    tau = 10.0
    dt = 1.0
    ComplementaryLIFNeuronState(0.0, 0.0, tau, 1.0, dt, exp(-dt / tau))
end

_finite(value::Float64) = isfinite(value)

function _validated_alpha(s::ComplementaryLIFNeuronState)
    if !_finite(s.tau) || s.tau <= 0.0 || !_finite(s.dt) || s.dt <= 0.0
        error("tau and dt must be positive")
    end
    ratio = -s.dt / s.tau
    alpha = ratio < -700.0 ? 0.0 : exp(ratio)
    if !_finite(alpha) || alpha < 0.0 || alpha >= 1.0
        error("alpha must be in [0, 1)")
    end
    return alpha
end

function validate(s::ComplementaryLIFNeuronState)
    if !_finite(s.v_pos) || !_finite(s.v_neg) || abs(s.v_pos) > V_MAX || abs(s.v_neg) > V_MAX
        return false
    end
    if !_finite(s.v_threshold) || s.v_threshold <= 0.0
        return false
    end
    try
        _validated_alpha(s)
    catch _e
        return false
    end
    return true
end

function step!(s::ComplementaryLIFNeuronState, I_ext::Float64=0.0; dt::Float64=s.dt)
    if !isfinite(I_ext) || !validate(s)
        return -2
    end
    alpha = _validated_alpha(s)
    inp_pos = max(I_ext, 0.0)
    inp_neg = max(-I_ext, 0.0)
    v_pos_next = alpha * s.v_pos + inp_pos
    v_neg_next = alpha * s.v_neg + inp_neg
    diff = v_pos_next - v_neg_next
    if !isfinite(v_pos_next) || !isfinite(v_neg_next) || !isfinite(diff) || abs(v_pos_next) > V_MAX || abs(v_neg_next) > V_MAX
        return -2
    end
    s.alpha = alpha
    if diff >= s.v_threshold
        s.v_pos = 0.0
        s.v_neg = 0.0
        return 1
    end
    if diff <= -s.v_threshold
        s.v_pos = 0.0
        s.v_neg = 0.0
        return -1
    end
    s.v_pos = v_pos_next
    s.v_neg = v_neg_next
    return 0
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = ComplementaryLIFNeuronState()
    s.dt = dt
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext)
        trace[t] = result == -2 ? NaN : s.v_pos - s.v_neg
        if result == 1 || result == -1
            spikes += 1
        end
    end
    return trace, spikes
end

end # module ClifAccel

File: src/sc_neurocore/accel/go/services/clif.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service for clif

package services

import (
	"errors"
	"math"
)

const clifVMax = 1.0e12

// ComplementaryLIFNeuronState holds the neuron state.
type ComplementaryLIFNeuronState struct {
	VPos       float64
	VNeg       float64
	Tau        float64
	VThreshold float64
	Dt         float64
	Alpha      float64
}

// NewComplementaryLIFNeuron creates a new ComplementaryLIFNeuron neuron with default parameters.
func NewComplementaryLIFNeuron() *ComplementaryLIFNeuronState {
	s := &ComplementaryLIFNeuronState{VPos: 0.0, VNeg: 0.0, Tau: 10.0, VThreshold: 1.0, Dt: 1.0}
	s.Alpha = math.Exp(-s.Dt / s.Tau)
	return s
}

func clifFinite(value float64) bool { return !math.IsNaN(value) && !math.IsInf(value, 0) }

func (s *ComplementaryLIFNeuronState) validatedAlpha() (float64, error) {
	if !clifFinite(s.Tau) || s.Tau <= 0.0 {
		return 0, errors.New("tau must be positive")
	}
	if !clifFinite(s.Dt) || s.Dt <= 0.0 {
		return 0, errors.New("dt must be positive")
	}
	ratio := -s.Dt / s.Tau
	if ratio < -700.0 {
		return 0.0, nil
	}
	alpha := math.Exp(ratio)
	if !clifFinite(alpha) || alpha < 0.0 || alpha >= 1.0 {
		return 0, errors.New("alpha must be in [0, 1)")
	}
	return alpha, nil
}

func (s *ComplementaryLIFNeuronState) validate() (float64, error) {
	if !clifFinite(s.VPos) || !clifFinite(s.VNeg) || math.Abs(s.VPos) > clifVMax || math.Abs(s.VNeg) > clifVMax {
		return 0, errors.New("CLIF membrane paths outside safety envelope")
	}
	if !clifFinite(s.VThreshold) || s.VThreshold <= 0.0 {
		return 0, errors.New("threshold must be positive")
	}
	return s.validatedAlpha()
}

// Step advances the neuron by one timestep.
func (s *ComplementaryLIFNeuronState) Step(iExt float64) (int, error) {
	if !clifFinite(iExt) {
		return 0, errors.New("current must be finite")
	}
	alpha, err := s.validate()
	if err != nil {
		return 0, err
	}
	inpPos := math.Max(iExt, 0.0)
	inpNeg := math.Max(-iExt, 0.0)
	vPosNext := alpha*s.VPos + inpPos
	vNegNext := alpha*s.VNeg + inpNeg
	diff := vPosNext - vNegNext
	if !clifFinite(vPosNext) || !clifFinite(vNegNext) || !clifFinite(diff) || math.Abs(vPosNext) > clifVMax || math.Abs(vNegNext) > clifVMax {
		return 0, errors.New("CLIF membrane candidate outside safety envelope")
	}
	s.Alpha = alpha
	if diff >= s.VThreshold {
		s.VPos, s.VNeg = 0.0, 0.0
		return 1, nil
	}
	if diff <= -s.VThreshold {
		s.VPos, s.VNeg = 0.0, 0.0
		return -1, nil
	}
	s.VPos, s.VNeg = vPosNext, vNegNext
	return 0, nil
}

// SimulateComplementaryLIFNeuron runs the neuron for n steps.
func SimulateComplementaryLIFNeuron(nSteps int, iExt float64) ([]float64, int) {
	s := NewComplementaryLIFNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result, err := s.Step(iExt)
		if err != nil {
			trace[t] = math.NaN()
			continue
		}
		trace[t] = s.VPos - s.VNeg
		if result != 0 {
			spikes++
		}
	}
	return trace, spikes
}
